lesson_1.py

```diff
@@ -1,59 +1,80 @@
-    * ``name`` - a product's name, like apple, cheese etc.
-    * ``price`` - a price for a single unit
-#. ``Product`` instance should have a behavior of calculating total
-   price for a specified quantity of goods
-#. Implement ``ShoppingCart`` class
-#. ``ShoppingCart`` instance should combine products instances and
-    corresponding purchased quantities.
-#. ``ShoppingCart`` instance should implement a method to calculate
-   the total price of entire cart.
-
-"""
-import random
 from typing import Union
 
 
 class Product():
-    def __init__(self, name: str, price: int):
+    def __init__(self, name: str, price: float):
         self.name = name
         self.price = price
 
     def calculate_cost(self, quantity: Union[int, float]):
         return round(quantity * self.price, 2)
 
+    def __str__(self):
+        return self.name
+
+    def __float__(self):
+        return self.price
+
+    def __repr__(self):
+        return f'class-{self.__class__}, product name - {self.name}, price - {self.price}'
+
 
 class Shopping_Cart():
     def __init__(self):
         self.products: List[Product] = []
         self.quantities: List[Union[int, float]] = []
 
     def add_to_cart(self, product: Product, quantity: Union[int, float]):
-        self.products.append(product)
-        self.quantities.append(quantity)
+        if product not in self.products:
+            self.products.append(product)
+            self.quantities.append(quantity)
+        else:
+            idx = self.products.index(product)
+            self.quantities[idx] += quantity
 
     def show_total_cost(self):
         total = 0
         for product, quantity in zip(self.products, self.quantities):
             total += product.calculate_cost(quantity)
+            print(f'Product in cart now is {str(product)}, quantity is {quantity}')
         return total
 
+    def __add__(self, other):
+        new_cart = Shopping_Cart()
+        new_cart.products = self.products.copy()
+        new_cart.quantities = self.quantities.copy()
+        if isinstance(other, Product):
+            new_cart.add_to_cart(other, 1)
+        if isinstance(other, Shopping_Cart):
+            if other == new_cart:
+                new_cart.quantities = [quantity * 2 for quantity in new_cart.quantities]
+                return new_cart
+            for product, quantity in zip(other.products, other.quantities):
+                new_cart.add_to_cart(product, quantity)
+        return new_cart
 
-def main():
-    # goods = ['Cheese', 'Apple', 'Butter', 'Meat']
-    Cheese = Product('Chheae', 350)
-    Apple = Product('Apple', 42)
-    Meet = Product('Meet', 200)
-    cart = Shopping_Cart()
-    cart.add_to_cart(Cheese, 2)
-    cart.add_to_cart(Apple, 1.5)
-    cart.add_to_cart(Meet, 2.5)
-    print(cart.show_total_cost())
+    def __eq__(self, other):
+        return set(self.products) == set(other.products), set(self.quantities) == set(other.quantities)
```

Comment:
Повертаєш кортеж з двох `bool`, замість одного `True/False`.
Методи порівняння (`__eq__`, `__gt__`, `__lt__` etc.) повині повертати `bool` або `NotImplemented`.
```suggestion
        if not isinstance(other, ShoppingCart):
            return False
        self_content = {(product, quantity) for product, quantity in zip(self.products, self.quantities)}
        other_content = {(product, quantity) for product, quantity in zip(other.product, other.quantities)}
        return self_content == other_content
```

Reply:
Можна звісно, залишити як у тебе, але тоді треба от так:
```suggestion
        return set(self.products) == set(other.products) and set(self.quantities) == set(other.quantities)
```